dm/master/election.go

```diff
@@ -47,29 +59,14 @@ func (s *Server) electionNotify(ctx context.Context) {
 				s.leader = oneselfLeader
 				s.closeLeaderClient()
 				s.Unlock()
-			case election.RetireFromLeader, election.IsNotLeader:
-				if notify == election.RetireFromLeader {
-					s.scheduler.Close()
-				}
-
-				leader, leaderID, leaderAddr, err2 := s.election.LeaderInfo(ctx)
-				if err2 != nil {
-					log.L().Warn("get leader info", zap.Error(err2))
-					continue
-				}
-
-				if notify == election.RetireFromLeader {
-					log.L().Info("current member retire from the leader", zap.String("leader", leaderID), zap.String("current member", s.cfg.Name))
-				} else {
-					log.L().Info("get new leader", zap.String("leader", leaderID), zap.String("current member", s.cfg.Name))
-				}
+			} else {
+				// this member is not leader
+				log.L().Info("get new leader", zap.String("leader", leaderInfo.ID), zap.String("current member", s.cfg.Name))
 
 				s.Lock()
-				s.leader = leader
-				s.createLeaderClient(leaderAddr)
+				s.leader = leaderInfo.ID
+				s.createLeaderClient(leaderInfo.Addr)
```

Comment:
`s.Lock` is used to protect `s.createLeaderClient(leaderInfo.Addr)`, and no need to protect `s.leader`, right?

Reply:
also need protect `s.leader`, because need to get `s.leader` when call function `isLeaderAndNeedForward`

Reply:
Do we need to protect `s.leader = ""` in L41?

Reply:
yes, I add lock in a33eecb